plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -632,14 +634,74 @@ private CpsScript parseScript() throws IOException {
             trusted = new CpsGroovyShellFactory(this).forTrusted().build();
             shell = new CpsGroovyShellFactory(this).withParent(trusted).build();
 
-            s = (CpsScript) shell.reparse("WorkflowScript",script);
+            s = (CpsScript) shell.reparse("WorkflowScript", script);
 
             for (Entry<String, String> e : loadedScripts.entrySet()) {
                 shell.reparse(e.getKey(), e.getValue());
             }
-        } catch (RuntimeException | Error x) {
+        } catch (Exception x) {
+            // Suspected groovyjarjarasm.asm.MethodTooLargeException or a
+            // org.codehaus.groovy.control.MultipleCompilationErrorsException
+            // whose collection of errors refers to MethodTooLargeException.
+            // Per review comments, we do not want to statically compile a
+            // dependency on the groovyjarjarasm.asm.MethodTooLargeException
+            // internals, so gauge hitting it via String name comparisons.
+            // Other cases may be (subclasses of) RuntimeException or Error.
+            Exception mtlEx = null;
+            int ecCount = 0;
+
+            // Clean up first
             closeShells();
-            throw x;
+
+            if (x.getClass().getSimpleName().equals("MethodTooLargeException")) {
+                mtlEx = x;
+                ecCount = 1;
+            } else if (x instanceof MultipleCompilationErrorsException) {
+                ErrorCollector ec = ((MultipleCompilationErrorsException) x).getErrorCollector();
+                ecCount = ec.getErrorCount();
+
+                for (int i = 0; i < ecCount; i++) {
+                    Exception ex = ec.getException(i);
+                    if (ex == null)
+                        continue;
+
+                    LOGGER.log(Level.FINE, "Collected Exception #" + i + ": " + ex.toString());
+                    if (ex.getClass().getSimpleName().equals("MethodTooLargeException")) {
+                        mtlEx = ex;
+                        break;
+                    }
+                }
+            }
+
+            if (mtlEx == null || ecCount < 1) {
+                // Some other exception type, or collection did not include MTL, rethrow as-is
+                throw x;
+            }
+
+            String msg = "FAILED to parse WorkflowScript (the pipeline script) due to MethodTooLargeException";
+            if (ecCount > 1) {
+                msg += " (and other issues)";
+            }
+            // Short message suffices, not much that a pipeline developer
+            // can do with the stack trace into the guts of groovy
+            msg += "; please refactor to simplify code structure and/or move logic to a Jenkins Shared Library: ";
+            msg += mtlEx.getMessage();
+
+            // Make a note in server log
+            LOGGER.log(Level.SEVERE, msg);
+
```

Comment:
Please do not; this is up to the owner of the project to fix, not the server admin.

```suggestion
```

Reply:
After some consideration (and per comments below to another code block), I've restored this server log message - but at a "FINER" level so it is normally no-op. Still, can be an aid to Jenkins admins (e.g. developers who would reproduce the problem on their local instance).

Reply:
Later will push a revision which uses a dedicated logger (snatching the idea from "CpsFlowExecution.timing" sub-logger) for easier one-off enablement of just this logic.